File: banner/src/main/java/com/qfxl/view/indicator/BaseIndicator.java
/*
 * Copyright (c) 2018 Frank Xu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.qfxl.view.indicator;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.widget.LinearLayout;

import com.qfxl.view.banner.BannerView;


/**
 * <pre>
 *     author : qfxl
 *     e-mail : xuyonghong0822@gmail.com
 *     time   : 2017/07/19
 *     desc   : 基础指示器
 *     version: 1.0
 * </pre>
 */
public abstract class BaseIndicator extends LinearLayout implements ViewPager.OnPageChangeListener {
    /**
     * 关联的LoopViewPager
     */
    protected BannerView mBannerView;
    /**
     * item的总个数
     */
    protected int itemTotalCount;
    /**
     * ViewPager item选中监听
     */
    private OnPageSelectListener mOnPageSelectListener;

    public interface OnPageSelectListener {
        /**
         * ViewPager item选中监听
         *
         * @param position
         */
        void onItemSelected(int position);
    }


    public BaseIndicator(Context context) {
        this(context, null);
    }

    public BaseIndicator(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BaseIndicator(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setOrientation(HORIZONTAL);
        setGravity(Gravity.CENTER);
    }

    /**
     * 设置ViewPager的滑动监听
     *
     * @param onPageSelectListener
     */
    public void setOnPageSelectListener(OnPageSelectListener onPageSelectListener) {
        mOnPageSelectListener = onPageSelectListener;
    }

    /**
     * 关联LoopViewPager
     *
     * @param bannerView
     */
    public void setViewPager(BannerView bannerView) {
        if (bannerView == mBannerView) {
            return;
        }
        if (bannerView != null && bannerView.getAdapter() != null) {
            mBannerView = bannerView;
            mBannerView.addOnPageChangeListener(this);
            itemTotalCount = mBannerView.getDecorAdapter().getRealCount();
            //只有数量>1才允许设置指示器
            if (itemTotalCount > 1) {
                createIndicators(itemTotalCount);
            }
        }
    }

    /**
     * 获取item的总个数
     *
     * @return
     */
    public int getItemTotalCount() {
        return itemTotalCount;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        onItemScrolled(position % mBannerView.getDecorAdapter().getRealCount(), positionOffset, positionOffsetPixels);
    }

    @Override
    public void onPageSelected(int position) {
        if (mOnPageSelectListener != null) {
            mOnPageSelectListener.onItemSelected(mBannerView.getCurrentRealItem());
        }
        onItemSelected(mBannerView.getCurrentRealItem());
    }

    /**
     * 选中item
     *
     * @param position 当前被选中item的位置
     */
    protected abstract void onItemSelected(int position);

    /**
     * 创建指示器
     *
     * @param itemCount item个数
     */
    protected abstract void createIndicators(int itemCount);

    /**
     * 监听item滑动的距离
     *
     * @param position
     * @param positionOffset
     * @param positionOffsetPixels
     */
    protected void onItemScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }
}

File: banner/src/main/kotlin/com.github.banner/indicator/BaseIndicator.kt
/*
 *
 *  * Copyright (C)  XU YONGHONG, Open Source Project
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *     http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *
 *
 */

package com.github.banner.indicator

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.github.banner.Banner
import com.github.banner.callback.OnBannerPageChangeCallback

/**
 * author : qfxl
 * e-mail : xuyonghong0822@gmail.com
 * time   : 2021/11/20
 * desc   :
 * version: 1.0
 */

abstract class BaseIndicator @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : View(context, attrs, defStyle), OnBannerPageChangeCallback {
    /**
     * gap between each indicator
     */
    var itemSpace = 0

    /**
     * width of each indicator
     */
    var itemWidth = 0

    /**
     * width of each selected indicator
     */
    var itemSelectWidth = itemWidth

    /**
     * height of each indicator
     */
    var itemHeight = 0

    /**
     * height of each selected indicator
     */
    var itemSelectHeight = itemHeight

    /**
     * number of indicators, this number is the realCount of Banner#adapter
     */
    var itemCount = 0

    /**
     * default color of the indicator
     */
    var defaultColor = 0

    /**
     * selected color of the indicator
     */
    var selectedColor = 0

    /**
     * orientation associated with Banner
     */
    var orientation = Banner.HORIZONTAL

    /**
     * current page
     */
    var currentPage = 0

    /**
     * indicator offset
     */
    var offset = 0f

    /**
     * indicator offset pixels
     */
    var offsetPixels = 0

    /**
     * indicator paint
     */
    val indicatorPaint by lazy(LazyThreadSafetyMode.NONE) {
        Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = defaultColor
        }
    }

    init {
        require(orientation == Banner.HORIZONTAL || orientation == Banner.VERTICAL) { "Indicator orientation must be Banner.HORIZONTAL or Banner.VERTICAL" }
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        offset = positionOffset
        offsetPixels = positionOffsetPixels
    }

    override fun onPageSelected(position: Int) {
        currentPage = position
        invalidate()
    }


    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        if (orientation == Banner.HORIZONTAL) {
            drawHorizontal(canvas)
        } else {
            drawVertical(canvas)
        }
    }

    override fun onPageScrollStateChanged(state: Int) {

    }

    /**
     * Check if need to set default values
     */
    open fun setDefaultValue() {

    }

    /**
     * draw horizontal indicators
     * @param canvas
     */
    abstract fun drawHorizontal(canvas: Canvas?)

    /**
     * draw vertical indicators
     * @param canvas
     */
    abstract fun drawVertical(canvas: Canvas?)
}